### ISSUE TITLE
Boogie allows identifiers to start with . and @, which SMTLIB does not

### ISSUE THREAD

**Issue by zvonimir** [boogie-org/boogie#7]
The following legal Boogie program (based on its current language spec):
```
procedure foo() {
  var .x:int;
  assert .x == 0;
}
```
causes Boogie to throw an exception when used with CVC4:
```
boogie test.bpl /proverOpt:SOLVER=cvc4
```

Here is the exception excerpt:
```
Prover error: (error "Parse Error: <stdin>:11.15: cannot declare or define symbol `.x'; symbols starting with . and @ are reserved in SMT-LIB
```

**Comment by delcypher** [comment]
I'd like to disallow use of ``@`` in a variable name because of #5. Of course we could change the attribute set reference operator from ``@`` to something else if my proposal is accepted.

**Comment by zvonimir** [comment]
Disallowing @ solves only one half on this problem :). Breaking backwards compatibility also means that potentially many tools will have to adjust their Boogie output. We (meaning team SMACK) can adjust relatively easily, but I am not sure about others. If we disallow @ then we should probably disallow . as well.

**Comment by pdeligia** [comment]
Another way (as I mentioned in the duplicate issue #8) would be to sanitize the two prefixes inside Boogie. So we would have an initial pass that basically detects vars that start with @ or . and change them to another symbol. I think that I saw Boogie doing this already for one of the prefixes, but I might be wrong. This would also mean that an external tool would not have to change. I am not sure which is the best way to approach this though, and it could be related to what Dan proposed in #5 or not.

**Comment by zvonimir** [comment]
Yes, my preference would be to sanitize within Boogie, but I will not push strong in that direction since we can adapt our tool either way.

**Comment by delcypher** [comment]
I'm fine with sanitizing ``@`` and ``.`` for now but I wanted to bring up #5 because if it was implemented in it's current form ``@`` wouldn't be allowed in Boogie variable names anyway.

**Comment by delcypher** [comment]
Are you sure about ``@`` being allowed in identifier names? This doesn't parse

```
procedure foo() {
  var @x:int;
  assert @x == 0;
}
```

**Comment by zvonimir** [comment]
Ah, sorry, I think that @ sneaked into this since Corral generates internally variables starting with @. Sorry about that. See http://research.microsoft.com/en-us/um/people/leino/papers/krml178.pdf on page  4 where identifiers are defined and @ is indeed not allowed.

**Comment by pdeligia** [comment]
I think the main problem is with the . prefix, as SMACK generates quite a few constants (e.g. .str) using it.

**Comment by delcypher** [comment]
Hmm let me look into this. ``SMTLibNamer`` at a glance is supposed to disallow ``.`` in an identifier

```
  public class SMTLibNamer : UniqueNamer
  {
    // The following Boogie ID characters are not SMT ID characters: `'\#    
    const string idCharacters = "~!@$%^&*_-+=<>.?/";
```

but clearly it's not working. I will investigate...

**Comment by delcypher** [comment]
Hmm well I managed to get Boogie to emit ``|.x|``  (i.e. it is quoted) instead of ``.x`` but CVC4 still rejects this

```
(error "Parse Error: x.smt2:3.17: cannot declare or define symbol `.x'; symbols starting with . and @ are reserved in SMT-LIB

  (declare-fun |.x| () Int)
                ^
")
```

I'm not sure if this a bug in CVC4 or this really is illegal in SMT-LIB. I've e-mailed the SMT-LIB mailing list and we'll see what they say.

**Comment by delcypher** [comment]
The [mailing list post is here](http://www.cs.nyu.edu/pipermail/smt-lib/2015/000878.html)

**Comment by delcypher** [comment]
@zvonimir  @pdeligia
Okay I've tried to fix the issue in bed8d4a027c635a146911a2ce973046f02a8e719

Does this fix your problem?

**Comment by delcypher** [comment]
I'm closing this. Please re-open if it's still a problem.
